File: src/agent/slotFiller.ts
import { toTitleCase } from '@/utils/strings';
import type { RoutedIntent } from './types';

const WEEKDAY_ORDER: Record<string, number> = {
  sunday: 0,
  monday: 1,
  tuesday: 2,
  wednesday: 3,
  thursday: 4,
  friday: 5,
  saturday: 6,
};

const MONTHS = [
  'january',
  'february',
  'march',
  'april',
  'may',
  'june',
  'july',
  'august',
  'september',
  'october',
  'november',
  'december',
];

export interface SlotFillOptions {
  userTimeZone?: string;
  now?: Date;
}

interface ParsedDateParts {
  date?: Date;
  timeApplied?: boolean;
}

const clampValue = (value: number, max: number): number => {
  if (Number.isNaN(value)) return 0;
  if (value < 0) return 0;
  if (value > max) return max;
  return value;
};

const minutesToISO = (date: Date): string => {
  return date.toISOString();
};

const resolveNow = (now?: Date): Date => (now ? new Date(now) : new Date());

const findWeekday = (text: string, now: Date): ParsedDateParts => {
  const match = text.match(/\b(next|this)?\s*(sunday|monday|tuesday|wednesday|thursday|friday|saturday)\b/i);
  if (!match) return {};

  const [, hint, weekdayRaw] = match;
  if (!weekdayRaw) {
    return {};
  }
  const weekday = WEEKDAY_ORDER[weekdayRaw.toLowerCase()];
  if (weekday === undefined) {
    return {};
  }
  const today = new Date(now);
  const todayIdx = today.getDay();
  let daysAhead = (weekday - todayIdx + 7) % 7;

  if (daysAhead === 0 && hint?.toLowerCase() === 'next') {
    daysAhead = 7;
  } else if (daysAhead === 0 && hint?.toLowerCase() === 'this') {
    daysAhead = 0;
  } else if (daysAhead === 0 && !hint) {
    daysAhead = 7; // assume future if unspecified and same day mentioned
  }

  const target = new Date(today);
  target.setDate(today.getDate() + daysAhead);
  return { date: target };
};

const findRelativeDay = (text: string, now: Date): ParsedDateParts => {
  if (/\btoday\b/i.test(text)) {
    return { date: new Date(now) };
  }
  if (/\btomorrow\b/i.test(text)) {
    const tomorrow = new Date(now);
    tomorrow.setDate(tomorrow.getDate() + 1);
    return { date: tomorrow };
  }
  if (/\bday after tomorrow\b/i.test(text)) {
    const target = new Date(now);
    target.setDate(target.getDate() + 2);
    return { date: target };
  }
  return {};
};

const findExplicitDate = (text: string, now: Date): ParsedDateParts => {
  const isoMatch = text.match(/\b(\d{4})-(\d{2})-(\d{2})\b/);
  if (isoMatch) {
    const [, y, m, d] = isoMatch;
    const date = new Date(Number(y), Number(m) - 1, Number(d));
    return { date };
  }

  const slashMatch = text.match(/\b(\d{1,2})\/(\d{1,2})(?:\/(\d{2,4}))?\b/);
  if (slashMatch) {
    const [, monthRaw, dayRaw, yearRaw] = slashMatch;
    if (!monthRaw || !dayRaw) {
      return {};
    }
    const year = yearRaw ? Number(yearRaw.length === 2 ? `20${yearRaw}` : yearRaw) : now.getFullYear();
    const date = new Date(year, Number(monthRaw) - 1, Number(dayRaw));
    return { date };
  }

  const monthMatch = text.match(/\b(jan(?:uary)?|feb(?:ruary)?|mar(?:ch)?|apr(?:il)?|may|jun(?:e)?|jul(?:y)?|aug(?:ust)?|sep(?:tember)?|oct(?:ober)?|nov(?:ember)?|dec(?:ember)?)\s+(\d{1,2})(?:,?\s*(\d{4}))?\b/i);
  if (monthMatch) {
    const [, monthRaw, dayRaw, yearRaw] = monthMatch;
    if (!monthRaw || !dayRaw) {
      return {};
    }
    const month = MONTHS.indexOf(monthRaw.toLowerCase());
    if (month < 0) {
      return {};
    }
    const year = yearRaw ? Number(yearRaw) : now.getFullYear();
    const date = new Date(year, month, Number(dayRaw));
    return { date };
  }

  return {};
};

const applyTime = (text: string, seed: Date | undefined, now: Date): ParsedDateParts => {
  const timeMatch = text.match(/\b(?:at|@)?\s*(\d{1,2})(?::(\d{2}))?\s*(am|pm)?\b/i);
  if (!timeMatch) return {};
  const [, hourRaw, minuteRaw, meridiemRaw] = timeMatch;
  if (!hourRaw) {
    return {};
  }
  let hours = clampValue(Number(hourRaw), 23);
  const minutes = clampValue(Number(minuteRaw ? Number(minuteRaw) : 0), 59);

  if (meridiemRaw) {
    const meridiem = meridiemRaw.toLowerCase();
    if (meridiem === 'pm' && hours < 12) {
      hours += 12;
    }
    if (meridiem === 'am' && hours === 12) {
      hours = 0;
    }
  }

  const base = seed ? new Date(seed) : new Date(now);
  base.setHours(hours, minutes, 0, 0);
  return { date: base, timeApplied: true };
};

const ensurePastIsFuture = (date: Date, now: Date): Date => {
  if (date.getTime() < now.getTime()) {
    const adjusted = new Date(date);
    adjusted.setDate(adjusted.getDate() + 7);
    return adjusted;
  }
  return date;
};

const deriveSlots = (text: string, baseIntent: RoutedIntent, options?: SlotFillOptions): Record<string, string> => {
  const now = resolveNow(options?.now);
  const resultSlots: Record<string, string> = { ...baseIntent.slots };
  const normalizedLabel = (baseIntent.rawLabel || baseIntent.label).toLowerCase();

  const weekday = findWeekday(text, now);
  const relative = findRelativeDay(text, now);
  const explicit = findExplicitDate(text, now);

  let candidateDate = explicit.date ?? relative.date ?? weekday.date;

  if (!candidateDate && /\btonight\b/i.test(text)) {
    candidateDate = new Date(now);
    candidateDate.setHours(20, 0, 0, 0);
  }

  const time = applyTime(text, candidateDate ?? new Date(now), now);
  if (time.date) {
    candidateDate = time.date;
  }

  if (candidateDate) {
    const adjusted =
      baseIntent.label.toLowerCase().includes('schedule')
        ? ensurePastIsFuture(candidateDate, now)
        : candidateDate;
    const iso = minutesToISO(adjusted);
    if (normalizedLabel.includes('schedule')) {
      resultSlots.start = iso;
      if (!time.timeApplied) {
        const end = new Date(adjusted);
        end.setHours(end.getHours() + 1);
        resultSlots.end = minutesToISO(end);
      }
    } else if (normalizedLabel.includes('goal')) {
      const [datePart] = iso.split('T');
      if (datePart) {
        resultSlots.due = datePart;
      }
    } else if (normalizedLabel.includes('journal')) {
      resultSlots.ts = iso;
    }
  }

  const durationMatch = text.match(/for\s+(\d{1,2})\s*(minutes?|hours?)/i);
  if (durationMatch) {
    const [, qtyRaw, unitRaw] = durationMatch;
    if (qtyRaw) {
      const qty = Number(qtyRaw);
      const unit = unitRaw && unitRaw.toLowerCase().startsWith('hour') ? 60 : 1;
      const minutes = qty * unit;
      resultSlots.duration_minutes = String(minutes);
    }
  }

  if (!resultSlots.title && normalizedLabel.includes('goal')) {
    const match = text.match(/(?:goal|aim|plan)\s+(?:to|is to)?\s*(.+)$/i);
    if (match && match[1]) {
      resultSlots.title = toTitleCase(match[1].trim());
    }
  }

  if (!resultSlots.title && normalizedLabel.includes('journal')) {
    const titleMatch = text.match(/(?:about|on)\s+([^.!?]+)/i);
    if (titleMatch && titleMatch[1]) {
      resultSlots.title = toTitleCase(titleMatch[1].trim());
    }
  }

  return resultSlots;
};

export const SlotFiller = {
  fill(text: string, intent: RoutedIntent, options?: SlotFillOptions): RoutedIntent {
    const slots = deriveSlots(text, intent, options);
    return {
      ...intent,
      slots,
    };
  },
};

File: services/python-ai/app/agent/slot_filler.py
import re
from copy import deepcopy
from datetime import datetime, timedelta, timezone
from typing import Dict, Optional, TypedDict

from .types import RoutedIntent
from .utils.strings import to_title_case

WEEKDAY_ORDER = {
    "sunday": 0,
    "monday": 1,
    "tuesday": 2,
    "wednesday": 3,
    "thursday": 4,
    "friday": 5,
    "saturday": 6,
}

MONTHS = [
    "january",
    "february",
    "march",
    "april",
    "may",
    "june",
    "july",
    "august",
    "september",
    "october",
    "november",
    "december",
]


class SlotFillOptions(TypedDict, total=False):
    userTimeZone: str
    now: datetime


def _clamp_value(value: int, max_value: int) -> int:
    if not isinstance(value, int):
        return 0
    if value < 0:
        return 0
    if value > max_value:
        return max_value
    return value


def _minutes_to_iso(date: datetime) -> str:
    return date.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


def _resolve_now(now: Optional[datetime]) -> datetime:
    return datetime.fromisoformat(now.isoformat()) if now else datetime.now(timezone.utc)


def _find_weekday(text: str, now: datetime) -> Dict[str, datetime]:
    match = re.search(r"\b(next|this)?\s*(sunday|monday|tuesday|wednesday|thursday|friday|saturday)\b", text, flags=re.I)
    if not match:
        return {}
    hint, weekday_raw = match.groups()
    weekday_idx = WEEKDAY_ORDER.get((weekday_raw or "").lower())
    if weekday_idx is None:
        return {}
    today_idx = now.weekday() if hasattr(now, "weekday") else datetime.now(timezone.utc).weekday()
    days_ahead = (weekday_idx - today_idx + 7) % 7
    if days_ahead == 0 and hint and hint.lower() == "next":
        days_ahead = 7
    elif days_ahead == 0 and hint and hint.lower() == "this":
        days_ahead = 0
    elif days_ahead == 0 and not hint:
        days_ahead = 7
    target = now + timedelta(days=days_ahead)
    return {"date": target}


def _find_relative_day(text: str, now: datetime) -> Dict[str, datetime]:
    if re.search(r"\btoday\b", text, flags=re.I):
        return {"date": now}
    if re.search(r"\btomorrow\b", text, flags=re.I):
        return {"date": now + timedelta(days=1)}
    if re.search(r"\bday after tomorrow\b", text, flags=re.I):
        return {"date": now + timedelta(days=2)}
    return {}


def _find_explicit_date(text: str, now: datetime) -> Dict[str, datetime]:
    iso_match = re.search(r"\b(\d{4})-(\d{2})-(\d{2})\b", text)
    if iso_match:
        year, month, day = iso_match.groups()
        return {"date": datetime(int(year), int(month), int(day), tzinfo=timezone.utc)}
    slash_match = re.search(r"\b(\d{1,2})\/(\d{1,2})(?:\/(\d{2,4}))?\b", text)
    if slash_match:
        month_raw, day_raw, year_raw = slash_match.groups()
        year_val = int(f"20{year_raw}") if year_raw and len(year_raw) == 2 else int(year_raw or now.year)
        return {"date": datetime(year_val, int(month_raw), int(day_raw), tzinfo=timezone.utc)}
    month_match = re.search(r"\b(jan(?:uary)?|feb(?:ruary)?|mar(?:ch)?|apr(?:il)?|may|jun(?:e)?|jul(?:y)?|aug(?:ust)?|sep(?:tember)?|oct(?:ober)?|nov(?:ember)?|dec(?:ember)?)\s+(\d{1,2})(?:,?\s*(\d{4}))?\b", text, flags=re.I)
    if month_match:
        month_raw, day_raw, year_raw = month_match.groups()
        month_idx = MONTHS.index(month_raw.lower()) if month_raw else -1
        if month_idx >= 0:
            year_val = int(year_raw) if year_raw else now.year
            return {"date": datetime(year_val, month_idx + 1, int(day_raw), tzinfo=timezone.utc)}
    return {}


def _apply_time(text: str, seed: Optional[datetime], now: datetime) -> Dict[str, datetime | bool]:
    match = re.search(r"\b(?:at|@)?\s*(\d{1,2})(?::(\d{2}))?\s*(am|pm)?\b", text, flags=re.I)
    if not match:
        return {}
    hour_raw, minute_raw, meridiem_raw = match.groups()
    hours = _clamp_value(int(hour_raw), 23)
    minutes = _clamp_value(int(minute_raw) if minute_raw else 0, 59)
    if meridiem_raw:
        meridiem = meridiem_raw.lower()
        if meridiem == "pm" and hours < 12:
            hours += 12
        if meridiem == "am" and hours == 12:
            hours = 0
    base = datetime.fromisoformat(seed.isoformat()) if seed else now
    base = base.replace(hour=hours, minute=minutes, second=0, microsecond=0)
    return {"date": base, "timeApplied": True}


def _ensure_past_is_future(date: datetime, now: datetime) -> datetime:
    if date < now:
        return date + timedelta(days=7)
    return date


def _derive_slots(text: str, base_intent: RoutedIntent, options: Optional[SlotFillOptions]) -> Dict[str, str]:
    now = _resolve_now(options.get("now") if options else None)
    result_slots: Dict[str, str] = deepcopy(base_intent.slots)
    normalized_label = (base_intent.rawLabel or base_intent.label).lower()

    weekday = _find_weekday(text, now)
    relative = _find_relative_day(text, now)
    explicit = _find_explicit_date(text, now)
    candidate_date = explicit.get("date") or relative.get("date") or weekday.get("date")

    if not candidate_date and re.search(r"\btonight\b", text, flags=re.I):
        candidate_date = now.replace(hour=20, minute=0, second=0, microsecond=0)

    time_applied = _apply_time(text, candidate_date or now, now)
    if time_applied.get("date"):
        candidate_date = time_applied["date"]  # type: ignore[assignment]

    if candidate_date:
        adjusted = _ensure_past_is_future(candidate_date, now) if "schedule" in base_intent.label.lower() else candidate_date
        iso = _minutes_to_iso(adjusted)
        if "schedule" in normalized_label:
            result_slots["start"] = iso
            if not time_applied.get("timeApplied"):
                end = adjusted + timedelta(hours=1)
                result_slots["end"] = _minutes_to_iso(end)
        elif "goal" in normalized_label:
            date_part = iso.split("T")[0]
            if date_part:
                result_slots["due"] = date_part
        elif "journal" in normalized_label:
            result_slots["ts"] = iso

    duration_match = re.search(r"for\s+(\d{1,2})\s*(minutes?|hours?)", text, flags=re.I)
    if duration_match:
        qty_raw, unit_raw = duration_match.groups()
        qty = int(qty_raw)
        unit = 60 if unit_raw and unit_raw.lower().startswith("hour") else 1
        minutes = qty * unit
        result_slots["duration_minutes"] = str(minutes)

    if "title" not in result_slots and "goal" in normalized_label:
        match = re.search(r"(?:goal|aim|plan)\s+(?:to|is to)?\s*(.+)$", text, flags=re.I)
        if match and match.group(1):
            result_slots["title"] = to_title_case(match.group(1).strip())

    if "title" not in result_slots and "journal" in normalized_label:
        title_match = re.search(r"(?:about|on)\s+([^.!?]+)", text, flags=re.I)
        if title_match and title_match.group(1):
            result_slots["title"] = to_title_case(title_match.group(1).strip())

    return result_slots


def fill(text: str, intent: RoutedIntent, options: Optional[SlotFillOptions] = None) -> RoutedIntent:
    slots = _derive_slots(text, intent, options)
    return RoutedIntent(**{**intent.model_dump(), "slots": slots})
